Source/FieldSolver/ImplicitSolvers/ThetaImplicitEM.cpp

```diff
@@ -0,0 +1,213 @@
+        "theta parameter for theta implicit time solver must be between 0.5 and 1.0");
+
+    std::string nlsolver_type_str;
+    pp.query("nonlinear_solver", nlsolver_type_str);
+    if (nlsolver_type_str=="picard") {
+        m_nlsolver_type = NonlinearSolverType::Picard;
+        m_max_particle_iterations = 1;
+        m_particle_tolerance = 0.0;
+    }
+    else if (nlsolver_type_str=="newton") {
+        m_nlsolver_type = NonlinearSolverType::Newton;
+        pp.query("max_particle_iterations", m_max_particle_iterations);
+        pp.query("particle_tolerance", m_particle_tolerance);
+    }
+    else {
+        WARPX_ABORT_WITH_MESSAGE(
+            "invalid nonlinear_solver specified. Valid options are picard and newton.");
+    }
+
+    // Define the nonlinear solver
+    if (m_nlsolver_type == NonlinearSolverType::Picard) {
+        m_nlsolver = std::make_unique<PicardSolver<WarpXSolverVec,ThetaImplicitEM>>();
+        m_nlsolver->Define(m_E, this);
+    }
+    else if (m_nlsolver_type == NonlinearSolverType::Newton) {
+        m_nlsolver = std::make_unique<NewtonSolver<WarpXSolverVec,ThetaImplicitEM>>();
+        m_nlsolver->Define(m_E, this);
+    }
+
+    m_is_defined = true;
+}
+
+void ThetaImplicitEM::PrintParameters () const
+{
+    if (!m_WarpX->Verbose()) { return; }
+    amrex::Print() << std::endl;
+    amrex::Print() << "-----------------------------------------------------------" << std::endl;
+    amrex::Print() << "----------- THETA IMPLICIT EM SOLVER PARAMETERS -----------" << std::endl;
+    amrex::Print() << "-----------------------------------------------------------" << std::endl;
+    amrex::Print() << "Time-bias parameter theta:  " << m_theta << std::endl;
+    amrex::Print() << "max particle iterations:    " << m_max_particle_iterations << std::endl;
+    amrex::Print() << "particle tolerance:         " << m_particle_tolerance << std::endl;
+    if (m_nlsolver_type==NonlinearSolverType::Picard) {
+        amrex::Print() << "Nonlinear solver type:      Picard" << std::endl;
+    }
+    else if (m_nlsolver_type==NonlinearSolverType::Newton) {
+        amrex::Print() << "Nonlinear solver type:      Newton" << std::endl;
+    }
+    m_nlsolver->PrintParams();
+    amrex::Print() << "-----------------------------------------------------------" << std::endl;
+    amrex::Print() << std::endl;
+}
+
+void ThetaImplicitEM::OneStep ( const amrex::Real  a_time,
+                                const amrex::Real  a_dt,
+                                const int          a_step )
+{
+    using namespace amrex::literals;
+    amrex::ignore_unused(a_step);
+
+    // Fields have E^{n} and B^{n}
+    // Particles have p^{n} and x^{n}.
+
+    // Save the values at the start of the time step,
+    m_WarpX->SaveParticlesAtImplicitStepStart ( );
+
+    // Save the fields at the start of the step
+    m_Eold.Copy( m_WarpX->getMultiLevelField(FieldType::Efield_fp) );
+    m_E = m_Eold; // initial guess for E
```

Comment:
We may want to remove the copy operator, to avoid that future developers accidentally copy large MultiFab without realizing it. (This is the approach that AMReX takes in general for the MultiFab object)
```suggestion
    m_E.Copy( m_WarpX->getMultiLevelField(FieldType::Efield_fp) ); // initial guess for E
```

Reply:
Not sure I understand. Copy is a standard and expected operator for the WarpXSolverVec object. This operator is used in both Picard and Newton methods.